### PR TITLE
search: hint type:commit or type:diff if query contains commit params

### LINKED ISSUE #12527: Author searches report 'no results', should hint to use type:commit or type:diff
Some filters result in 'No result' and do not help the user understand why. Example:

1. Search repo:sourcgraph/sourcegraph author:rob@sourcegraph.com
2. Note the user has not included type commit or diff 
3. Run search

#### Expected behavior:
A tip is provided that author searches require type:commit or type:diff.

#### Actual behavior:
3. Receive the "No Results" banner

Identify combinations of languages and types which are not compatible and add hints for the user as to how to produce a valid result.

### PR DESCRIPTION
**Pull request by rvantonder**
Fixes #12527 @rrhyne LMK if you'd like a different wording.

Screen Shot 2020-08-24 at 5 47 56 PM

### CODE REVIEW

**Comment by rrhyne** [comment]
I would make it more clinical: 

"Your query contains the field 'author' which requires type:commit or type:diff"

Possibly add "in the query' if you don't think that is clear enough.

This change will be helpful, thanks!

File: internal/search/query/validate.go [hunk 1]
```diff
@@ -349,6 +349,25 @@ func validateRepoRevPair(nodes []Node) error {
...
+	VisitParameter(nodes, func(field, value string, _ bool, _ Annotation) {
+		if field == FieldAuthor || field == FieldBefore || field == FieldAfter || field == FieldMessage {
+			seenCommitParam = field
+		}
+		if field == FieldType && (value == "commit" || value == "diff") {
+			typeCommitExists = true
+		}
+	})
+	if seenCommitParam != "" && !typeCommitExists {
+		return fmt.Errorf(`your query contains the field "%s", so please also add either type:commit or type:diff to the query`, seenCommitParam)
```
Comment (rvantonder) [inline internal/search/query/validate.go:366]:
```suggestion
		return fmt.Errorf(`your query contains the field '%s', which requires type:commit or type:diff in the query`, seenCommitParam)
```

File: internal/search/query/validate_test.go [hunk 2]
```diff
@@ -72,6 +72,10 @@ func TestAndOrQuery_Validation(t *testing.T) {
 			input: "repo:foo@a rev:b",
 			want:  "invalid syntax. You specified both @ and rev: for a repo: filter and I don't know how to interpret this. Remove either @ or rev: and try again",
 		},
+		{
+			input: "repo:foo author:rob@saucegraph.com",
+			want:  `your query contains the field "author", so please also add either type:commit or type:diff to the query`,
```
Comment (rvantonder) [inline internal/search/query/validate_test.go:77]:
```suggestion
			want:  `your query contains the field 'author', which requires type:commit or type:diff in the query`,
```

**Review by keegancsmith — APPROVED** [review]

**Comment by limitedmage** [comment]
**suggestion**: can we add a search suggestion button to the alert for adding the type?

**Comment by rvantonder** [comment]
> suggestion: can we add a search suggestion button to the alert for adding the type?

I started off this way in a [separate PR](https://github.com/sourcegraph/sourcegraph/pull/12577) but ultimately think it's low value / janky in the current context. The proposed queries is not really compatible with the new parser/tree. I want us to rework query and alert suggestions (brought this up before, when we/I are ready we can dig into it). The existing proposed queries aren't clear/obvious, they're cluttered with `patterntype:blahblah`, they're problematic for search expressions, and the code for converting parse errors to alerts isn't great. So I'm deciding to avoid adding more of that until we can rework things.

**Comment by rvantonder** [comment]
> suggestion: can we add a search suggestion button to the alert for adding the type?

I started off this way in a [separate PR](https://github.com/sourcegraph/sourcegraph/pull/12577) but ultimately think it's low value / janky in the current context. The proposed queries is not really compatible with the new parser/tree. I want us to rework query and alert suggestions (brought this up before, when we/I are ready we can dig into it). The existing proposed queries aren't clear/obvious, they're cluttered with `patterntype:blahblah`, they're problematic for search expressions, and the code for converting parse errors to alerts isn't great. So I'm deciding to avoid adding more of that until we can rework things.

**Pull request merged**